### PR TITLE
Support custom python-executable

### PR DESCRIPTION
**Pull request by MaratFM**
**Problem**: build pipelines and infrastructure at big companies often targets multiple independent python projects. Currently, in order to use pip-sync, pip-tools have to be installed in every environment. When pip-compile does not have such restriction.

**Solution**: add the ability to pass custom python executable.  When the option provided, list of installed packages will be collected from the target environment and all install/uninstall commands will be executed inside that environment. 

This PR implements a proof concept for collecting early feedback on the idea and implementation. If concept and task will be accepted, sufficient test coverage will be added and PR re-submitted.

Thank you!

**Added "--python-executable" option for pip-sync command. It allows targeting environments other than current**: 

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by MaratFM** [comment]
@atugushev could you please comment if this PR makes sense and maintainers agree to add such an option

**Comment by atugushev** [comment]
Hello @MaratFM,

Tganks for pinging! I believe that this functionality would help for such issues as #1087. So i'm :+1: on this.

File: piptools/scripts/sync.py
Comment (atugushev) [inline piptools/scripts/sync.py:59]: We could use `type=click.Path(exists=True)` instead of this check below:

```python
    if python_executable is not None and not os.path.exists(python_executable):
         msg = "Python executable {} not found"
         log.error(msg.format(python_executable))
         sys.exit(2)
```

Comment (MaratFM) [reply piptools/scripts/sync.py:59]: Great idea, thanks!

File: piptools/scripts/sync.py
Comment (atugushev) [inline piptools/scripts/sync.py:59]: Subjective: maybe `--python` option (like in `virtualenv`)?

Comment (MaratFM) [reply piptools/scripts/sync.py:59]: No problem changing. Interesting, which name is the most commonly used. I saw `--python-executable` in mypy
https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-python-executable

Comment (MaratFM) [reply piptools/scripts/sync.py:59]: Overall, it seems like with `--python` there are more ways for misinterpretation, so I am leaning towards more explicit `--python-executable` but the final call is yours

File: piptools/scripts/sync.py
Comment (atugushev) [inline piptools/scripts/sync.py:136]: I would rewrite this to:

```suggestion
    paths = None if python_executable is None else get_sys_path_for_python_executable(python_executable) 
```

or vice versa

File: piptools/scripts/sync.py
Comment (atugushev) [reply piptools/scripts/sync.py:59]: > Overall, it seems like with `--python` there are more ways for misinterpretation, so I am leaning towards more explicit `--python-executable`

Oh, that's interesting. Let's wait for other opinions.

**Review by atugushev — COMMENTED** [review]
Generally, it looks good. One thing to consider: could we improve `--python-executable` to pass binary name somehow? For example: `pip-sync --python-executable=python3`.

**Comment by MaratFM** [comment]
> Generally, it looks good. One thing to consider: could we improve `--python-executable` to pass binary name somehow? For example: `pip-sync --python-executable=python3`.

Great idea, I can try using `shutil.which` for resolving
https://docs.python.org/3/library/shutil.html?highlight=.which#shutil.which

Looks like it works well for aliases and full paths, so we can always use it for resolving:
```py
In [1]: import shutil

In [2]: import sys

In [3]: shutil.which('python')
Out[3]: 'D:\\Envs\\test\\python.EXE'

In [4]: shutil.which(sys.executable)
Out[4]: 'd:\\envs\\test\\python.exe'
```

**Comment by MaratFM** [comment]
@atugushev please take a look when you will have a chance. I added support for binary executable name (like `pip-sync --python-executable=python3`). In addition, while testing locally, I found that when we use a custom python executable it is no guaranteed that pip version is compatible. So I added an explicit check for it.

**Comment by MaratFM** [comment]
@atugushev gentle ping

**Comment by MaratFM** [comment]
@webknjaz @jdufresne could you please take a look if you have time 🙂

File: piptools/scripts/sync.py
Comment (webknjaz) [inline piptools/scripts/sync.py:13]: Have you checked if `packaging` is available directly, maybe as a transitive dep?

File: piptools/scripts/sync.py
Comment (webknjaz) [inline piptools/scripts/sync.py:27]: Instead of hardcoding such things, it's possible to parse it from the metadata via `importlib_metadata` or `pkg_util` or smth like that.

Comment (webknjaz) [reply piptools/scripts/sync.py:27]: You may find some inspiration here https://github.com/cherrypy/cherrypy/blob/e00d162/docs/conf.py#L32-L57

File: piptools/scripts/sync.py
Comment (MaratFM) [reply piptools/scripts/sync.py]: It is available, but because other parts of code are using imports from `pip._vendor`, mixing with direct imports will break `isinstance` like checks. Not sure about the original motivation to use  `pip._vendor` imports, most probably it is to make it more stable and not rely on libraries except pip

File: piptools/scripts/sync.py
Comment (MaratFM) [reply piptools/scripts/sync.py]: thanks for the suggestion, updated to use `get_distribution` and compare the version with the specifier

**Comment by MaratFM** [comment]
@webknjaz @atugushev could you please take a look

**Comment by MaratFM** [comment]
@jdufresne could you please help with the review?

**Comment by jdufresne** [comment]
Sorry, I will not have the time to review this. Please carry on with the other reviewers.

File: piptools/scripts/sync.py
Comment (webknjaz) [inline piptools/scripts/sync.py:138]: FWIW it's recommended (although controversial sometimes) to let the logging module take care of rendering things. And this is why it's best to pass a separate string with percent-placeholders as the first arg and pass the vars separately, instead of pre-rendering that string.
This also allows different tooling to aggregate log entries coming from the same place but with different args.

File: piptools/scripts/sync.py
Comment (webknjaz) [inline piptools/scripts/sync.py:140]: This block looks like it could belong in a dedicated function.

File: tests/test_cli_sync.py
Comment (webknjaz) [inline tests/test_cli_sync.py:262]: PEP 257 requires sentences (with a period at the end)
```suggestion
    Make sure sync command can run with `--python-executable` option.
```

**Comment by webknjaz** [comment]
Same here. I hope @atugushev will be able to get this to the finish line.

File: piptools/scripts/sync.py
Comment (MaratFM) [reply piptools/scripts/sync.py]: Thanks, I updated logging calls. 

BTW, I was just following the pattern used in the file, it could be a separate discussion if all log statements should be aligned and switch to logging formatting.

File: piptools/scripts/sync.py
Comment (MaratFM) [reply piptools/scripts/sync.py]: thanks, extracted _validate_python_executable method

File: tests/test_cli_sync.py
Comment (MaratFM) [reply tests/test_cli_sync.py:252]: sounds good, will keep aligned with the rest of the code

File: tests/test_cli_sync.py
Comment (MaratFM) [reply tests/test_cli_sync.py]: updated here and in other places I found in my code

**Review by atugushev — COMMENTED** [review]
I've caught a trace when a given python executable is not found:

```
$ pip-sync -a --python-executable=python3.9
Traceback (most recent call last):
  File "/Users/albert/Projects/pip-tools/.venv/bin/pip-sync", line 11, in <module>
    load_entry_point('pip-tools', 'console_scripts', 'pip-sync')()
  File "/Users/albert/Projects/pip-tools/.venv/lib/python3.8/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/albert/Projects/pip-tools/.venv/lib/python3.8/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/albert/Projects/pip-tools/.venv/lib/python3.8/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/albert/Projects/pip-tools/.venv/lib/python3.8/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/albert/Projects/pip-tools/piptools/scripts/sync.py", line 126, in cli
    _validate_python_executable(python_executable)
  File "/Users/albert/Projects/pip-tools/piptools/scripts/sync.py", line 200, in _validate_python_executable
    log.error(msg, python_executable, pip_version, required_pip_specification)
  File "/Users/albert/Projects/pip-tools/piptools/logging.py", line 40, in error
    self.log(message, *args, **kwargs)
  File "/Users/albert/Projects/pip-tools/piptools/logging.py", line 24, in log
    click.secho(prefix + message, *args, **kwargs)
TypeError: secho() got multiple values for argument 'err'
```

**Comment by atugushev** [comment]
> I reverted logging message passing, was confused by the recommendation to delegate message formatting to the logging system ([#1333 (comment)](https://github.com/jazzband/pip-tools/pull/1333#discussion_r647266617)). Seems like pip-tools uses a custom logging wrapper that has incompatible method contracts.

Thanks @MaratFM! Let's address that in a following up PR.

File: tests/test_cli_sync.py
Comment (atugushev) [inline tests/test_cli_sync.py:251]: There are too much of monkeypatches. Ideally, there should be only `@mock.patch("piptools.sync.run")`.

File: piptools/scripts/sync.py
Comment (atugushev) [reply piptools/scripts/sync.py:68]: I'll resolve this. Let it be `--python-executable`.

File: piptools/utils.py
Comment (atugushev) [inline piptools/utils.py:373]: It seems the `None` branch left uncovered. Options:
1. add a test
2. add pragma marker to skip this branch

File: tests/test_utils.py
Comment (atugushev) [inline tests/test_utils.py:441]: `assert result` should be checked to make sure that below assert would be reached.

Comment (MaratFM) [reply tests/test_utils.py:442]: good point, added an assert

File: piptools/utils.py
Comment (MaratFM) [reply piptools/utils.py]: added pragma marker

File: tests/test_cli_sync.py
Comment (MaratFM) [reply tests/test_cli_sync.py]: refactored

File: piptools/utils.py
Comment (atugushev) [inline piptools/utils.py:375]:
```suggestion
    )  # pragma: no branch
```

There is a [no branch](https://coverage.readthedocs.io/en/coverage-4.3.3/branch.html#structurally-partial-branches) pragma.

**Comment by MaratFM** [comment]
@atugushev seems like with "# pragma: no branch" it stills shows partial coverage, so I reverted back to "# pragma: no cover"

File: piptools/utils.py
Comment (atugushev) [inline piptools/utils.py:375]:
```suggestion
    requirement = next(  # pragma: no branch
        (r for r in project_dist.requires() if r.name == "pip"), None
    )
```

This should work.

**Comment by MaratFM** [comment]
@atugushev sorry for changing tests back and forth, I could not run tests for all these platforms locally and have to wait when someone will trigger the GitHub actions pipeline

**Review by atugushev — APPROVED** [review]
Thanks @MaratFM for your patience. Awesome work! 👍🏻

**Pull request merged**
